Fixes #4011

matthiaskraaz opened the pull request:
I took the same approach as https://github.com/hashicorp/consul/issues/8510 was solved:

- removed the -m option from useradd so it does not try to create the directory
- added /var/go to directories to be created by package

chadlwilson commented:
Thx!


Did you happen to check 
- whether this has any difference in effect for `.deb`s on Ubuntu or Debian? IIRC the same script is used for both - I naively assume that the SELinux problems might be similar on both platforms.
- whether similar change needs to be made in `InstallerTypeAgent.groovy` (believe it does, as I imagine the agent installation has the same problem and `before-install.sh.erb` is used for both agent and server packages)

matthiaskraaz commented:
Oops. Let me check...

matthiaskraaz commented:
I added InstallerTypeAgent.groovy to the PR and also did a lot of testing.

### Ubuntu 22.04

I didn't succeed in enabling SELinux. 
According to https://askubuntu.com/questions/481293/selinux-implementation-in-ubuntu SELinux on Ubuntu is dead.
Is it worth investing more time to get an Ubuntu with SELinux running?
I tested my changes without SELinux.

- install agent
- connect agent to different host
- create and run pipeline
- install server
- connect agent to local server
- create and run pipeline
- fresh install
- install server
- connect agent from different host
- create and run pipeline
- install agent locally
- connect local agent to server

Result: changed deb works fine under Ubuntu 22.04 without SELinux

### Debian 12 without SELinux

- fresh install
- installed changed server deb
- installed git
- connected agent from different host
- created and run pipeline
- installed changed agent deb locally
- connected agent to server
- run pipeline
- fresh install
- installed changed agent dep
- installed git
- connect agent to different host
- created and run pipeline
- installed changed server deb locally
- connected agent to server
- run pipeline

Result: changed deb works fine under Debian 12 without SELinux

### Debian 12 with SELinux

Fresh install and activated SELinux then:
# apt update
# apt install selinux-basics selinux-policy-default auditd
# selinux-activate
# reboot
# check-selinux-installation
# setenforce 1
(Source: https://wiki.debian.org/SELinux/Setup)

Server fails to start:

```
INFO   | jvm ver. | 2024/04/07 15:03:06 | Error occurred during initialization of VM
INFO   | jvm ver. | 2024/04/07 15:03:06 | Failed to mark memory page as executable - check if grsecurity/PaX is enabled
ERROR  | wrapper  | 2024/04/07 15:03:06 | Failed to parse the version of Java. Resolving to the lowest supported version (1.4).
```

Tried also with official go-server_23.5.0-18179_all.deb. Server fails to start again with above error.

What do you think?

chadlwilson commented:
Wow, thanks for the detail here - that's far above and beyond!

If there are other problems with Debian SELinux (looks like the Tanuki wrapper's Java/JAVA_HOME detection isn't working for some reason we don't have to solve them here. Once we sort this our for RHEL, we can open a separate issue and see if anyone complains/cares. Part of me has been thinking about dropping Debian containers and raw Debian installs for official support anyway, as I don't hear much noise about it.

Even RHEL 9 / CentOS Stream 9 has issues right now due to #10722 and needs workarounds.

Anyway, couple of other things I'd like to check 
- I think the default **directory perms** are different when created by `useradd` vs here. I believe normally the group (`go`) has no default read permissions to the created home dir, i.e `0700` (or `cat /etc/login.defs | grep HOME_MODE` rather than `0750`. Whether this matters I am not sure, I can't recall if anything in particular is stored here by default (I think the relevant stuff is in `/etc/go`), or whether it's just for scripts/shells/ssh credentials etc. Am I missing something here? Do you think we should use something more restrictive by default?
- Do you anticipate if it makes any difference to **removal** of the package? I believe the current approach will leave `/var/go` behind because the directory is not package-owned. I don't know if it's a pro/con for package/removal to have this folder automatically cleaned up. Currently most important stuff is in `/etc/go` which **is cleaned up**, so probably doesn't matter to also clean up this dir and is maybe a pro. But if people have put SSH keys there for, e.g ssh git access, we would remove them on uninstall I suppose.

matthiaskraaz commented:
Let me check, what the old and new behavior of rpm and deb is regarding permissions of the dir. I suggest to keep the old behavior.

I believe that files not owned by the package are not removed even if in a directory owned by the package. Uninstall may give a warning though. But let me check.

chadlwilson commented:
To safe you some empirical checking, `0700` is the default on CentOS Stream 9 Minimal coming from `cat /etc/login.defs | grep HOME_MODE`, but I didn't check RHEL or earlier versions. To some extent the previous behaviour is dependent on distro-default, `umask`s etc I understand so we may not be able to match all, but `0700` seems sensible for a home dir, even with go using a dedicated `group` generally (and thus low risk of group readability being a problem).

When I checked the uninstall on CentOS Stream 9 with `23.5.0`'s rpm it did remove `/etc/go` entirely however I hadn't actually created files outside the package/started the server etc, so that might not have been a valid test. (excuse my ignorance, the intriciacies of rpms/debs etc is not my strong suit!)

Separately to this, I am impressed you seem to have managed to get a build going locally. If you had any major friction or feedback on the process of doing so, or the stuff at https://developer.gocd.org/current/ please raise issues/complaints/feedback and I'll fix it. Generally getting to the point of getting the installer builds going with `fpm` etc locally is even more painful than the rest of the build, and there are no specific instructions for linux where that stuff is actually easier, so good for you!

matthiaskraaz commented:
Thank you. And thank you for keeping GoCD alive.

Regarding the permissions: I think explicit `0700` has a very, very slight chance of breaking something somewhere, explicit `0750` has a very, very slight chance of being a security problem. Because it only matters if someone has actually added a user to the go group. Please make the call. You have the experience in that matter.

### Ubuntu 22.04 and Debian 12

apt purge:
- outputs: `dpkg: warning: while removing go-agent, directory '/var/go' not empty so not removed`
- keeps directory /var/go and file within

apt remove: 
- no warning
- keeps directory /var/go and file within

### CentOS 9

`rpm -e`:
- keeps directory /var/go and file within

I don't think rpm has a "purge" option.

### Build

I used WSL with Ubuntu. I had to

```
sudo apt install openjdk-17-jdk-headless
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh | bash
source ~/.bashrc
nvm install --lts
corepack enable
sudo apt install ruby-rubygems
sudo gem install fpm
sudo apt install rpm
```

I also had to 

`./gradlew --stop`

because nvm fiddling with the environment doesn't propagate to a running daemon.

I tried Java 21 first, but that gave me endlessly `waiting for isolate _isolateMain to check in` when trying to build.

matthiaskraaz commented:
OTOH, even the chrony package under CentOS removes /var/lib/chrony, the home directory of the chrony user.

chadlwilson commented:
OK, let's stick with this, should be fine I think!

> ### Build
> 
> I used WSL with Ubuntu. I had to
> 
> ```
> sudo apt install openjdk-17-jdk-headless
> curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh | bash
> source ~/.bashrc
> nvm install --lts
> corepack enable
> sudo apt install ruby-rubygems
> sudo gem install fpm
> sudo apt install rpm

Yeah, that looks about right, that's essentially what the build agents for GoCD itself do to pre-install.

https://github.com/gocd-contrib/gocd-oss-cookbooks/blob/a4dd63c7673963aac1c6a4023b764a70ed00400b/provision/provision-centos.sh#L148-L162

> ```
> 
> I also had to
> 
> `./gradlew --stop`
> 
> because nvm fiddling with the environment doesn't propagate to a running daemon.

Yeah, that's Gradle-life. Urggh.

> I tried Java 21 first, but that gave me endlessly `waiting for isolate _isolateMain to check in` when trying to build.

Yeah, I haven't validated either the build or GoCD itself with Java 21 yet, but that's a weird one - I wonder where that error is coming from, as it looks like a Dart error, and I can't think of a Java 21 interaction with Dart like that.... (if I recall correctly, only the sass compiler forked within the builds is built with Dart)

chadlwilson commented:
For what it's worth, the main build pipeline does run some rpm/deb driven installer tests which dont run on pull requests (due them needing to be post-signing and such), so when it has flowed through I'll let you know. It's the `installer-tests` pipeline at https://build.gocd.org/go (login with the `guest` button if you're interested)

If you feel like it, you can play with the experimental builds too :-)

matthiaskraaz commented:
My PR is just building. How exciting! :-)

chadlwilson commented:
_Experimental_ builds from https://www.gocd.org/download/#redhat `24.1.0-18170`+ have your fix in them; the installer tests were fine (it only tests a fresh install and an upgrade on CentOS Stream 8 of the RHEL adjacent distros)

`dnf install -y go-server-24.1.0-18710`

https://build.gocd.org/go/tab/build/detail/installer-tests/2410/install-tests/1/fresh-install-runInstance-4
https://build.gocd.org/go/tab/build/detail/installer-tests/2410/install-tests/1/upgrade-runInstance-4

chadlwilson commented:
Frontend isn't my strong suit either.

I'm not sure, depends whether you're looking to fix a bug, implement a feature, or experiment with finding an upgrade path for some technical debt/outdated libraries/tools?

It's not the most critical thing, but along the lines of this one, if you know something about rpms and debs and have an idea how we might approach #11354 at reasonable cost with the current repository infrastructure we have, that might be interesting?

matthiaskraaz commented:
LOL. I had this in my notes, because I tried to reproduce #4011 on an ARM instance first: 
_go-server-23.5.0-18179.noarch.rpm. says noarch. but contains x86-64 jre._

I'm on it.
